process-util-java9/src/main/java/org/jenkinsci/utils/process/ProcessUtils.java

```diff
@@ -0,0 +1,20 @@
+package org.jenkinsci.utils.process;
+
+import java.lang.reflect.Field;
+
+/**
+ * @author Kohsuke Kawaguchi
+ */
+@SuppressWarnings("Since15")
+public class ProcessUtils {
+    /**
+     * Figure out the UNIX process ID from {@link Process}
+     */
+    public static int getPid(Process p) {
+        try {
+            return (int) p.pid();
```

Comment:
All of this build-time complexity for _that_? Instead of

```java
try {
    return (int) p.getMethod("pid").invoke(p);
} catch (…some clauses here…) {
    try {
        Field f = p.getClass().getDeclaredField("pid");
        f.setAccessible(true);
        return (int) f.get(p);
    } catch (…) {
    }
     return -1;
}

Reply:
@jglick you can't use `Process.getDeclaredField("pid")` on java 9+ without a warning as the JVM start, and this will even just break with a future java release. 
introspection on java.* classes is now forbidden, we will have to live with this legitimate restriction (unfortunately this comes very late in java history)

Reply:
also, this build time complexity allows a cleaner code base imho. I hate to see such a 'try-then-retry-then..' logic in code trying to adapt to runtime

Reply:
> you can't use `Process.getDeclaredField("pid")` on java 9+

You would not need to—that code would be run only on Java 8. On 9+, you would be using the `pid()` method, which is `public` so it should be OK to access reflectively IIUC.

> this build time complexity allows a cleaner code base

My point is that it is far better to have a little complexity in plain old Java code, which people can read and debug pretty easily, than in build system complexity, which requires a lot more background to understand.

Reply:
I fully disagree with this pov. I find the maintenance effort way simpler with a cleaner codebase.